Dockerfile

```diff
@@ -9,6 +9,7 @@ COPY ./package-lock.json .
 COPY ./swagger.config.json .
 COPY ./tsconfig.json .
 
+RUN npm install
```

Comment:
Why do we need this?

Reply:
CI failed on two tries on two different days during step 10 of running 'docker build -t mofb-api .'  I am not super familiar with Docker but it appeared to be because there was no npm module in the Docker image to support the swaggerGen call in the 'npm run swagger' step.  Here is a snippet of the failure:

```
Step 10/14 : RUN npm run build
 ---> Running in 3f3e541c17a7
> mofb-api@2.0.0 build /app
> npm run swagger
> mofb-api@2.0.0 swagger /app
> swaggerGen -c ./swagger.config.json
sh: swaggerGen: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
```

I also noticed that 'npm-install' was removed from the swagger build script in package.json in PR #92 commit https://github.com/SCODEMeetup/mofb-api/pull/92/commits/41c6e69d56eb32b5552762212db0a7cf8a5d6946.  I thought that made sense for local development but also wondered if 'npm install' needed to be added for the docker build step. When I pushed that commit, CI worked.   

For more detail, here is a link to one of the failures (hopefully you can view it).  https://travis-ci.org/github/SCODEMeetup/mofb-api/builds/749106216

Reply:
So in LN 6 above we are copying the node modules to the docker image. So not really sure if npm install is needed as it would just generate the node modules which are already on the machine. 

Will take a look at your logs to see if there was another reason travis failed.

Reply:
So looking at the build. I think it is failing because it is pulling down a new image. So my guess is that the previous image had node_modules already installed automatically. Where I think this change should go is in file .travis.yml

add this as the first entry after **script:** so that install is done on every build and does not impact docker.

Reply:
Also probably better to use `npm ci` instead of `npm install`

Reply:
Agreed - `npm ci` should be used in any CI environment. But yea that's what I was thinking was that the `node_modules` directory was already being copied. Adding to `.travis.yml` sounds like a good next step to try. It also looks like `npm ci` was ran in #92 's travis ci build so I'm wondering where that went...

Reply:
> add this (npm ci) as the first entry after script: so that install is done on every build and does not impact docker.

I am wondering if it makes sense to place 'npm ci' as the first entry after before_install: because docker build calls 'npm run build' which is where swaggerGen is invoked.  I guess that depends on whether before_install: runs  before scripts.

Reply:
If you look at the build log for this PR it called `npm install` _and_ `npm ci`...which is not ideal. I do think that maybe docker needs to have an `npm install` call though since it calls `npm run build` which used to also call `npm install`. So I think it has been calling it twice all along. It's probably fine to have it run twice (I'm not a docker expert or else I would probably know of a better solution than this 😅 ).

Reply:
Good catch!  I guess 'npm ci' is implicitly run with the default install step.  I knew ```run npm build``` called ```npm install``` previously which is what gave me the idea of adding it to the Dockerfile.  The only question I have is whether I should change ```npm install``` to ```npm ci``` in Dockerfile to be consistent?

Reply:
No, since the dockerfile is responsible for preparing the environment in production, it will need to run the post install steps which `npm ci` does not run.

Reply:
👆 I may be wrong on that 😅 but `npm ci` does delete `node_modules` and reinstall everything which takes a while. I'm curious on input from @mihiramin89 though.

Reply:
> ```npm ci``` does delete node_modules and reinstall everything which takes a while

I think that is true.  I found this somewhat helpful: 

https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#dependency-management

travis-ci does ```npm ci``` when it detects a package-lock.json file.  They recommend putting an npm install command (or npm ci) at the top of the before_install: section so I just did a push with that and removed the ```npm install`` from Dockerfile.  I am curious if travis-ci will run ```npm ci``` a second time.

Reply:
> No, since the dockerfile is responsible for preparing the environment in production, it will need to run the post install steps which npm ci does not run.

Sorry, missed this earlier.  I am not sure what is meant by post install steps here?

Reply:
The last push passed CI but it ran ```npm ci``` twice.  

![Screen Shot 2020-12-14 at 3 41 20 PM](https://user-images.githubusercontent.com/31697968/102133297-e400cd00-3e22-11eb-94bf-814693c667e8.png)

https://travis-ci.org/github/SCODEMeetup/mofb-api/builds/749679083

```npm i``` in Dockerfile versus ```npm ci``` in .travis.yml

Pros:  

- Conforms with README Docker build instructions.
- Decouples from Docker build script using ```COPY .node_modules node_modules com```

Cons:
- ```npm ci``` too expensive?  ~10s

I think the pros may outweigh the cons (if a concern).  Could try adding ```install: skip``` to .travis.yml to see if that removes the second ```npm ci``` without causing problems with running the script.

Reply:
latest push has only one ```npm ci``` and CI passed, shaving 10s off time.

Reply:
Saw there was questions around `npm ci` vs `npm install`. I wanted to share what they have on NPM's [site](https://docs.npmjs.com/cli/v6/commands/npm-ci). 
<img width="751" alt="npm-ci___npm_Docs" src="https://user-images.githubusercontent.com/2163914/102161619-a2d5e080-3e55-11eb-8a49-759476ca4267.png">

Reply:
I am looking to see what calls we make when we deploy.. it should be the calls in the `cloudbuild.yaml` but I am seeing other calls as well when in google console. Based on the above, it looks like `npm ci` is what we should use since we have a `package-lock.json` and that should be the source of truth.

Reply:
Since we have this in the `cloudbuild.yaml` file https://github.com/SCODEMeetup/mofb-api/blob/ab0519740dc91c57f4d4adc476862dcd2a0a2882/cloudbuild.yaml#L4-L5 (and based off of Mihir's feedback) I think we should change that to use `npm ci` instead and then remove any `npm install/ci`'s from the dockerfile. Travis ci will install with `npm ci` automatically.

Reply:
@mycarrysun,  This is basically the approach I took with my commit late yesterday afternoon - with a twist.  I removed the ```npm install``` step from Dockerfile.  The travis CI install step was running ```npm ci```automatically, but it needed to be run explicitly in the before_install step for the Dockerfile to copy the node_modules in and allow ```npm run build``` to work.  

However, this meant ```npm ci``` was being run twice, so I skipped the install step which did nothing but run ```npm ci``` again for no reason.  Skipping the install step removed 10s from the Travis CI.  See snippet below.  Please review this code from latest commit.

```
before_install:
  - npm ci
  - docker build -t mofb-api .	  - docker build -t mofb-api .


install: skip

script:	script:
  - npm run tsc	  - npm run tsc
  - npm run lint	  - npm run lint
```

Reply:
Since docker doesn't need `npm ci` we don't need it in the `before_install` step - I think we should just do the regular order that travis expects (and automatically runs `npm ci`). There's no need to change the order.

Reply:
We do need to change the `cloudbuild.yaml` file still though

Reply:
>  Since docker doesn't need npm ci we don't need it in the before_install step - I think we should just do the regular order that travis expects (and automatically runs npm ci). There's no need to change the order.

From my experience this is not correct.  The cause of the original CI error was that docker build needed the ```npm install``` command  in the ```npm run build``` script which was removed in PR #92 .  The docker build was failing because the ```npm ci``` runs automatically in the default ```install:``` travis-ci step which occurs after the docker build is run .

By placing ```npm ci``` explicitly in the top of the ```before_install:``` step, the docker build and CI succeeds.  But even then, without skipping the install step, ```npm ci``` runs again automatically and unnecessarily, ergo the  reason ```install:skip```.  

Another option might be moving the docker build to an explicit ```install: step```.   The only reason I can think of why the docker build is in the ```before_install:``` step is because it was counting on the ```npm install``` that was previously in the. ```npm run build``` script, which it isn't any longer.  

I don't know what needs to change in cloudbuild.yaml.

Reply:
@mihiramin89, seeking input on the following:

1.  What needs to change in cloudbuild.yaml if anything?  Does GCP build using docker?  Does it do an npm ci before starting?  Sorry, this process is opaque to me.
2. You approved my changes on Monday.  Based on discussion between Mike and I above, do you have any second thoughts?  I am sure what I have done works for CI but would be willing to move the docker build from the before_install: step to the install: step if you think that would be better.

If you don't have any thoughts here I am almost inclined to merge to master and see if the deploy works or not and go from there.  😊

Reply:
The only thing that needs changed in cloudbuild.yaml is changing it from `npm install` to `npm ci`

Reply:
With the last commit ```npm install``` is out of the Dockerfile.  OK, so you are saying change:

args: ['install']  to

args: ['ci'] 

Makes sense and pretty easy to change back.  @mihiramin89  do you agree?

Reply:
yup i think that is all we need to do. I did not see any `npm ci` call in the log output when building in GCP. At this point, I think we should make the change and see if it works.

Reply:
Just curious about something.  Does GCP use docker for this app?

Reply:
@mihiramin89, @mycarrysun.  I will need a final review on this to merge to master.